spark/src/main/scala/com/acervera/osm4scala/spark/OsmPbfFormat.scala

```diff
@@ -0,0 +1,78 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package com.acervera.osm4scala.spark
+
+import java.net.URI
+
+import com.acervera.osm4scala.EntityIterator
+import com.acervera.osm4scala.spark.OsmPbfRowIterator._
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileStatus, Path}
+import org.apache.hadoop.mapreduce.Job
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.datasources.{FileFormat, OutputWriterFactory, PartitionedFile}
+import org.apache.spark.sql.sources.{DataSourceRegister, Filter}
+import org.apache.spark.sql.types._
+import org.apache.spark.util.SerializableConfiguration
+
+class OsmPbfFormat extends FileFormat with DataSourceRegister {
+
+  override def shortName(): String = "osm.pbf"
+
+  override def inferSchema(sparkSession: SparkSession,
+                           options: Map[String, String],
+                           files: Seq[FileStatus]): Option[StructType] = Some(OsmSqlEntity.schema)
+
+  override def prepareWrite(sparkSession: SparkSession,
+                            job: Job,
+                            options: Map[String, String],
+                            dataSchema: StructType): OutputWriterFactory =
+    throw new UnsupportedOperationException(
+      s"write is not supported for spark-osm-pbf files. If you need it, please create a issue and try to support the project."
+    )
+
+  override def isSplitable(sparkSession: SparkSession, options: Map[String, String], path: Path): Boolean = false
+
+  override protected def buildReader(sparkSession: SparkSession,
+                                     dataSchema: StructType,
+                                     partitionSchema: StructType,
+                                     requiredSchema: StructType,
+                                     filters: Seq[Filter],
+                                     options: Map[String, String],
+                                     hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
+
+    val broadcastedHadoopConf = sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    (file: PartitionedFile) =>
+      {
+        val path = new Path(new URI(file.filePath))
+        val fs = path.getFileSystem(broadcastedHadoopConf.value.value)
+        val status = fs.getFileStatus(path)
+        EntityIterator.fromPbf(fs.open(status.getPath)).toInternalRowIterator()
```

Comment:
Cool. So currently based on pbf structure, it reads these entities sequentially and put them into each partition?

Reply:
Yes, it is. I used it with real datafiles and it works fine. It is creating an iterator, so it is not going to be allocated in memory until Spark requires it.
If you have more than one file, this solution is good enough, because will create one partition per file.

I need to add a few testing and I will release a version.
Next step, the fun part. Trying to split the file to create more than one partition per file.